dex/networks/dcr/script.go

```diff
@@ -701,85 +508,64 @@ func DataPrefixSize(data []byte) uint8 {
-	if err != nil {
-		return nil, fmt.Errorf("error parsing utxo script addresses: %w", err)
+	// P2SH
+	if len(redeemScript) == 0 {
+		return nil, fmt.Errorf("no redeem script provided for P2SH pubkey script")
 	}
-	if nonStandard {
+
+	var redeemScriptType ScriptType
+	redeemScriptType, scriptAddrs = ExtractScriptAddrs(version, redeemScript, chainParams)
+	if redeemScriptType == ScriptUnsupported {
 		return &SpendInfo{
 			// SigScriptSize cannot be determined, leave zero.
 			ScriptAddrs:       scriptAddrs,
-			ScriptType:        scriptType,
-			NonStandardScript: true,
+			ScriptType:        scriptType, // still P2SH
+			NonStandardScript: true,       // but non-standard redeem script (e.g. contract)
 		}, nil
 	}
 
-	// Get the size of the signature script.
-	sigScriptSize := P2PKHSigScriptSize
-	// If it's a P2SH, the size must be calculated based on other factors.
-	if scriptType.IsP2SH() {
-		// Start with the signatures.
-		sigScriptSize = 74 * scriptAddrs.NRequired // 73 max for sig, 1 for push code
-		// If there are pubkey-hash addresses, they'll need pubkeys.
-		if scriptAddrs.NumPKH > 0 {
-			sigScriptSize += scriptAddrs.NRequired * (pubkeyLength + 1)
-		}
-		// Then add the length of the script and another push opcode byte.
-		sigScriptSize += len(redeemScript) + int(DataPrefixSize(redeemScript)) // push data length op code might be >1 byte
-	} else if !scriptType.IsP2PKH() {
-		// Some defensive checks here.
-		if scriptType.IsMultiSig() {
-			// Should have been caught by ParseScriptType.
-			return nil, fmt.Errorf("bare multisig is unsupported")
-		}
-		// P2PK also should have been caught by ParseScriptType.
-		return nil, fmt.Errorf("unsupported pkScript: %d", scriptType)
+	// NOTE: scriptAddrs are now from the redeemScript, and we add the multisig
+	// bit to scriptType.
+	if redeemScriptType.IsMultiSig() {
+		scriptType |= ScriptMultiSig
+	}
+
+	// If it's a P2SH, with a standard redeem script, compute the sigScript size
+	// for the expected number of signatures, pubkeys, and the redeem script
+	// itself. Start with the signatures.
+	sigScriptSize := (1 + DERSigLength) * scriptAddrs.NRequired // 73 max for sig, 1 for push code
+	// If there are pubkey-hash addresses, they'll need pubkeys.
+	if len(scriptAddrs.PkHashes) > 0 {
+		sigScriptSize += scriptAddrs.NRequired * (pubkeyLength + 1)
 	}
+	// Then add the length of the script and the push opcode byte(s).
+	sigScriptSize += len(redeemScript) + int(DataPrefixSize(redeemScript)) // push data length op code might be >1 byte
 	return &SpendInfo{
 		SigScriptSize: uint32(sigScriptSize),
 		ScriptAddrs:   scriptAddrs,
 		ScriptType:    scriptType,
 	}, nil
 }
 
-// ExtractContractHash extracts the contract P2SH address from a pkScript. A
-// non-nil error is returned if the hexadecimal encoding of the pkScript is
-// invalid, or if the pkScript is not a P2SH script.
-func ExtractContractHash(scriptHex string) ([]byte, error) {
```

Comment:
This wasn't used?

Reply:
Found no reference outside of tests.  It might have been at one point.